src/pow.cpp

```diff
@@ -89,3 +89,15 @@ bool CheckProofOfWork(uint256 hash, unsigned int nBits, const Consensus::Params&
 
     return true;
 }
+
+bool MaybeGenerateProof(const Consensus::Params& params, CBlockHeader* pblock, uint64_t& max_tries)
+{
+    static const int nInnerLoopCount = 0x10000;
+    uint256 blockHash = pblock->GetHash();
+    while (max_tries > 0 && pblock->nNonce < nInnerLoopCount && !CheckProofOfWork(blockHash, pblock->nBits, params)) {
+        ++pblock->nNonce;
+        blockHash = pblock->GetHash();
+        --max_tries;
+    }
+    return CheckProofOfWork(blockHash, pblock->nBits, params);
```

Comment:
It seems unnecessary to call `CheckProofOfWork` twice. You can move it from the while loop into the body, and make it return true immediately when it succeeds.